xdsl/dialects/csl_/csl_stencil.py

```diff
@@ -0,0 +1,271 @@
+class ExchangeDeclarationAttr(ParametrizedAttribute):
+    """
+    A simplified version of dmp.exchange, from which it should be lowered
+    """
+
+    name = "csl_stencil.exchange"
+
+    neighbor_param: ParameterDef[builtin.DenseArrayBase]
+
+    def __init__(
+        self,
+        neighbor: Sequence[int] | builtin.DenseArrayBase,
+    ):
+        data_type = builtin.i64
+        super().__init__(
+            [
+                (
+                    neighbor
+                    if isinstance(neighbor, builtin.DenseArrayBase)
+                    else builtin.DenseArrayBase.from_list(data_type, neighbor)
+                ),
+            ]
+        )
+
+    @classmethod
+    def from_dmp_exch_decl_attr(cls, src: dmp.ExchangeDeclarationAttr):
+        return cls(src.neighbor)
+
+    @property
+    def neighbor(self) -> tuple[int, ...]:
+        data = self.neighbor_param.as_tuple()
+        assert isa(data, tuple[int, ...])
+        return data
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string(f"<to {list(self.neighbor)}>")
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
+        parser.parse_characters("<")
+        parser.parse_characters("to")
+        to = parser.parse_comma_separated_list(
+            parser.Delimiter.SQUARE, parser.parse_integer
+        )
+        parser.parse_characters(">")
+
+        return [builtin.DenseArrayBase.from_list(builtin.i64, to)]
+
+
+@irdl_op_definition
+class PrefetchOp(IRDLOperation):
+    """
+    An op to indicate buffer prefetches.
+    """
+
+    name = "csl_stencil.prefetch"
+
+    input_stencil: Operand = operand_def(
+        stencil.TempType[Attribute] | memref.MemRefType[Attribute]
+    )
+
+    swaps: builtin.ArrayAttr[ExchangeDeclarationAttr] | None = opt_prop_def(
+        builtin.ArrayAttr[ExchangeDeclarationAttr]
+    )
+
+    size = prop_def(IntegerAttr[IntegerType])
+
+    topo: dmp.RankTopoAttr | None = opt_prop_def(dmp.RankTopoAttr)
+
+    result = result_def(memref.MemRefType)
+
+    def __init__(
+        self,
+        input_stencil: SSAValue | Operation,
+        size: IntegerAttr[IntegerType],
+        topo: dmp.RankTopoAttr | None = None,
+        swaps: Sequence[ExchangeDeclarationAttr] = [],
+        result_type: memref.MemRefType[Attribute] | None = None,
+    ):
+        super().__init__(
+            operands=[input_stencil],
+            properties={"size": size, "topo": topo, "swaps": builtin.ArrayAttr(swaps)},
+            result_types=[result_type],
+        )
+
+
+@irdl_op_definition
+class AccessOp(IRDLOperation):
+    """
+    A CSL stencil access that operates on data prefetched by `csl_stencil.prefetch`
+    """
+
+    T = Annotated[Attribute, ConstraintVar("T")]
+
+    name = "csl_stencil.access"
+    op: Operand = operand_def(memref.MemRefType)
+    offset: stencil.IndexAttr = attr_def(stencil.IndexAttr)
+    offset_mapping = opt_attr_def(stencil.IndexAttr)
```

Comment:
Is there a reason why we are using attributes instead of properties here?

Reply:
Just because that's what `stencil.access` does. Happy to change it though

Reply:
changed as suggested.